src/core/qgsvectorfilewriter.cpp

```diff
@@ -2550,16 +2552,30 @@ QgsVectorFileWriter::WriterError QgsVectorFileWriter::writeAsVectorFormat( Prepa
   int lastProgressReport = 0;
   long total = details.featureCount;
 
+  // Special rules for OGR layers
   if ( details.providerType == QLatin1String( "ogr" ) && !details.dataSourceUri.isEmpty() )
   {
     QStringList theURIParts = details.dataSourceUri.split( '|' );
-    QString srcFileName = theURIParts[0];
+    QgsStringMap srcUriParams;
+    QString srcFileName;
```

Comment:
Nice change! You can also get the filename from providerUriParams here and avoid this parsing